From 3ae23976862fa9036ab52e47c8a22025f5d76ca0 Mon Sep 17 00:00:00 2001
From: David Benjamin <davidben@google.com>
Date: Mon, 18 Mar 2024 22:43:46 +1000
Subject: [PATCH] Rework the test data story

We originally embedded test data because of deficiencies in Android's
build. Android had no way to specify test data with tests. That has
since been resolved, and the embedding mechanism has gotten unwieldy.

This unifies pki_test and crypto_test's test data story, and does so in
a way that all tests can participate in. (We can now use FileTest in
decrepit_test.)

Update-Note: This will require some tweaks to downstream builds. We no
longer emit an (unwieldy) crypto_test_data.cc file. Instead, tests will
expect test data be available at the current working directory. This can
be overridden with the BORINGSSL_TEST_DATA_ROOT environment variable.

Callers with more complex needs can build with
BORINGSSL_CUSTOM_GET_TEST_DATA and then link in an alternate
implementation of this function.

On the off chance some project needs it, I've kept the
embed_test_data.go script around for now, but I expect we can delete it
in the future.

Fixed: 681
Change-Id: If181ce043e1eea3148838f1bb4db9ee4bfda0d08
Reviewed-on: https://boringssl-review.googlesource.com/c/boringssl/+/67295
Commit-Queue: David Benjamin <davidben@google.com>
Reviewed-by: Bob Beck <bbe@google.com>
---
 CMakeLists.txt                 | 22 +--------------
 build.json                     |  6 ++--
 crypto/pkcs8/pkcs12_test.cc    |  3 +-
 crypto/test/file_test_gtest.cc |  2 +-
 crypto/test/test_data.cc       | 51 ++++++++++++++++++++++++++++++++++
 crypto/test/test_data.h        | 31 +++++++++++++++++++++
 crypto/x509/x509_test.cc       |  3 +-
 gen/sources.cmake              |  3 +-
 gen/sources.json               |  3 +-
 pki/test_helpers.cc            | 51 +++-------------------------------
 util/generate_build_files.py   | 34 +++++++----------------
 11 files changed, 106 insertions(+), 103 deletions(-)
 create mode 100644 crypto/test/test_data.cc
 create mode 100644 crypto/test/test_data.h

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 35fb5670df..1cff48f002 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -490,26 +490,6 @@ target_include_directories(
 # themselves as dependencies next to the target definition.
 add_custom_target(all_tests)
 
-# On Windows, CRYPTO_TEST_DATA is too long to fit in command-line limits.
-# TODO(davidben): CMake 3.12 has a list(JOIN) command. Use that when we've
-# updated the minimum version.
-set(EMBED_TEST_DATA_ARGS "")
-foreach(arg ${CRYPTO_TEST_DATA})
-  set(EMBED_TEST_DATA_ARGS "${EMBED_TEST_DATA_ARGS}${arg}\n")
-endforeach()
-file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/embed_test_data_args.txt"
-     "${EMBED_TEST_DATA_ARGS}")
-
-add_custom_command(
-  OUTPUT crypto_test_data.cc
-  COMMAND ${GO_EXECUTABLE} run util/embed_test_data.go -file-list
-          "${CMAKE_CURRENT_BINARY_DIR}/embed_test_data_args.txt" >
-          "${CMAKE_CURRENT_BINARY_DIR}/crypto_test_data.cc"
-  DEPENDS util/embed_test_data.go ${CRYPTO_TEST_DATA}
-  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
-
-add_library(crypto_test_data OBJECT crypto_test_data.cc)
-
 add_subdirectory(crypto)
 add_subdirectory(ssl/test)
 add_subdirectory(util/fipstools)
@@ -550,7 +530,7 @@ add_executable(urandom_test ${URANDOM_TEST_SOURCES})
 target_link_libraries(urandom_test test_support_lib boringssl_gtest crypto)
 add_dependencies(all_tests urandom_test)
 
-add_executable(crypto_test ${CRYPTO_TEST_SOURCES} $<TARGET_OBJECTS:crypto_test_data>)
+add_executable(crypto_test ${CRYPTO_TEST_SOURCES})
 target_link_libraries(crypto_test test_support_lib boringssl_gtest crypto)
 add_dependencies(all_tests crypto_test)
 
diff --git a/crypto/pkcs8/pkcs12_test.cc b/crypto/pkcs8/pkcs12_test.cc
index 459339d758..a210089158 100644
--- a/crypto/pkcs8/pkcs12_test.cc
+++ b/crypto/pkcs8/pkcs12_test.cc
@@ -25,11 +25,10 @@
 #include <openssl/stack.h>
 #include <openssl/x509.h>
 
+#include "../test/test_data.h"
 #include "../test/test_util.h"
 
 
-std::string GetTestData(const char *path);
-
 // kPassword is the password shared by most of the sample PKCS#12 files.
 static const char kPassword[] = "foo";
 
diff --git a/crypto/test/file_test_gtest.cc b/crypto/test/file_test_gtest.cc
index 2cb0896947..b1d35562a4 100644
--- a/crypto/test/file_test_gtest.cc
+++ b/crypto/test/file_test_gtest.cc
@@ -25,8 +25,8 @@
 
 #include <openssl/err.h>
 
+#include "test_data.h"
 
-std::string GetTestData(const char *path);
 
 class StringLineReader : public FileTest::LineReader {
  public:
diff --git a/crypto/test/test_data.cc b/crypto/test/test_data.cc
new file mode 100644
index 0000000000..9fe2aaaa06
--- /dev/null
+++ b/crypto/test/test_data.cc
@@ -0,0 +1,51 @@
+/* Copyright (c) 2024, Google Inc.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. */
+
+#include "test_data.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "file_util.h"
+
+#if !defined(BORINGSSL_CUSTOM_GET_TEST_DATA)
+std::string GetTestData(const char *path) {
+  const char *root = getenv("BORINGSSL_TEST_DATA_ROOT");
+  root = root != nullptr ? root : ".";
+
+  std::string full_path = root;
+  full_path.push_back('/');
+  full_path.append(path);
+
+  ScopedFILE file(fopen(full_path.c_str(), "rb"));
+  if (file == nullptr) {
+    fprintf(stderr, "Could not open '%s'.\n", full_path.c_str());
+    abort();
+  }
+
+  std::string ret;
+  for (;;) {
+    char buf[512];
+    size_t n = fread(buf, 1, sizeof(buf), file.get());
+    if (n == 0) {
+      if (feof(file.get())) {
+        return ret;
+      }
+      fprintf(stderr, "Error reading from '%s'.\n", full_path.c_str());
+      abort();
+    }
+    ret.append(buf, n);
+  }
+}
+#endif  // !BORINGSSL_CUSTOM_GET_TEST_DATA
diff --git a/crypto/test/test_data.h b/crypto/test/test_data.h
new file mode 100644
index 0000000000..b3f4b6ace3
--- /dev/null
+++ b/crypto/test/test_data.h
@@ -0,0 +1,31 @@
+/* Copyright (c) 2024, Google Inc.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. */
+
+#ifndef OPENSSL_HEADER_CRYPTO_TEST_TEST_DATA_H
+#define OPENSSL_HEADER_CRYPTO_TEST_TEST_DATA_H
+
+#include <string>
+
+// GetTestData returns the test data for |path|, or aborts on error. |path|
+// must be a slash-separated path, relative to the BoringSSL source tree. By
+// default, this is implemented by reading from the filesystem, relative to
+// the BORINGSSL_TEST_DATA_ROOT environment variable, or the current working
+// directory if unset.
+//
+// Callers with more complex needs can build with
+// BORINGSSL_CUSTOM_GET_TEST_DATA and then link in an alternate implementation
+// of this function.
+std::string GetTestData(const char *path);
+
+#endif  // OPENSSL_HEADER_CRYPTO_TEST_TEST_DATA_H
diff --git a/crypto/x509/x509_test.cc b/crypto/x509/x509_test.cc
index 4559b221b1..ffbcf7b861 100644
--- a/crypto/x509/x509_test.cc
+++ b/crypto/x509/x509_test.cc
@@ -35,6 +35,8 @@
 
 #include "internal.h"
 #include "../internal.h"
+#include "../test/file_util.h"
+#include "../test/test_data.h"
 #include "../test/test_util.h"
 #include "../x509v3/internal.h"
 
@@ -44,8 +46,6 @@
 #endif
 
 
-std::string GetTestData(const char *path);
-
 static const char kCrossSigningRootPEM[] = R"(
 -----BEGIN CERTIFICATE-----
 MIICcTCCAdqgAwIBAgIIagJHiPvE0MowDQYJKoZIhvcNAQELBQAwPDEaMBgGA1UE

diff --git a/pki/test_helpers.cc b/pki/test_helpers.cc
index b6712af515..490fba5aee 100644
--- a/pki/test_helpers.cc
+++ b/pki/test_helpers.cc
@@ -18,10 +18,13 @@
 #include "trust_store.h"
 #include "parser.h"
 #include <gtest/gtest.h>
+
 #include <openssl/bytestring.h>
 #include <openssl/mem.h>
 #include <openssl/pool.h>
 
+#include "../crypto/test/test_data.h"
+
 namespace bssl {
 
 namespace {
@@ -413,19 +416,7 @@
 }
 
 std::string ReadTestFileToString(const std::string& file_path_ascii) {
-  // Compute the full path, relative to the src/ directory.
-  fillins::FilePath src_root;
-  bssl::fillins::PathService::Get(fillins::BSSL_TEST_DATA_ROOT, &src_root);
-  fillins::FilePath filepath = src_root.AppendASCII(file_path_ascii);
-
-  // Read the full contents of the file.
-  std::string file_data;
-  if (!fillins::ReadFileToString(filepath, &file_data)) {
-    ADD_FAILURE() << "Couldn't read file: " << filepath.value();
-    return std::string();
-  }
-
-  return file_data;
+  return GetTestData(("pki/" + file_path_ascii).c_str());
 }
 
 void VerifyCertPathErrors(const std::string& expected_errors_str,
